File: src/org/jabby/blazenav/ChooseLetterAction.java
package org.jabby.blazenav;

import com.google.common.collect.Lists;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CustomShortcutSet;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.editor.Editor;

import javax.swing.*;
import java.util.*;

class ChooseLetterAction extends ActionBase {
    final Character key;
    final Collection<Integer> offsets;
    final Runnable unregisterShortcutsFn;
    final Runnable[] cleanupFns;

    ChooseLetterAction(Character key, Collection<Integer> offsets,
                       Runnable unregisterShortcutsFn, Runnable... cleanupFns) {
        this.key = key;
        this.offsets = offsets;
        this.unregisterShortcutsFn = unregisterShortcutsFn;
        this.cleanupFns = cleanupFns;
    }

    @Override public void actionPerformed(AnActionEvent e) {
        Editor editor = e.getData(PlatformDataKeys.EDITOR);

        unregisterShortcutsFn.run();

        if (editor != null) actionPerformed(editor);
    }

    void actionPerformed(Editor editor) {
        Map<Character, Integer> mappings = mapKeysToOffsets(offsets);
        JComponent editorComponent = editor.getContentComponent();

        if (mappings.size() > 1) {
            Runnable unhighlightCharsFn = highlightChars(editor, mappings);

            List<AnAction> nextActions = new ArrayList<AnAction>(mappings.size());
            Runnable unregisterShortcutsFn = createUnregisterShortcutsFn(editorComponent, nextActions);

            for (Character c : mappings.keySet()) {
                MoveCaretToOffsetAction next =
                        new MoveCaretToOffsetAction(mappings.get(c),
                                Utils.mergeVargs(cleanupFns, unregisterShortcutsFn, unhighlightCharsFn));
                next.registerCustomShortcutSet(new CustomShortcutSet(createKeyStroke(c)), editorComponent);
                nextActions.add(next);
            }

            setStatusText(editor, "n.candidates.found", mappings.size());
            setupCancellation(editorComponent,
                    Utils.mergeVargs(cleanupFns,
                            unregisterShortcutsFn, unhighlightCharsFn, createClearStatusFn(editor)));
        } else {
            assert mappings.size() == 1;
            setStatusText(editor, "one.candidate");
            moveCaretTo(editor, mappings.values().iterator().next());
            runAll(Utils.mergeVargs(cleanupFns, unregisterShortcutsFn));
        }
    }

    Map<Character, Integer> mapKeysToOffsets(Collection<Integer> offsets) {
        List<Integer> affectedOffsets =
                Lists.newArrayList(offsets).subList(0, Math.min(offsets.size(), SUBSTITUTIONS.length));
        Collections.sort(affectedOffsets);
        Map<Character, Integer> mappings = new HashMap<Character, Integer>(affectedOffsets.size());
        for (int i = 0; i < affectedOffsets.size(); i++) { mappings.put(SUBSTITUTIONS[i], affectedOffsets.get(i)); }
        return mappings;
    }
}

File: src/org/jabby/blazenav/ChooseLetterAction.kt
package org.jabby.blazenav

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.editor.Editor
import java.util.HashMap
import java.util.ArrayList
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.CustomShortcutSet

class ChooseLetterAction(val key: Char,
                         val offsets: Collection<Int>,
                         val unRegisterShortcutsFn: () -> Unit,
                         vararg val cleanupFns: () -> Unit): ActionBase() {

    public override fun actionPerformed(e: AnActionEvent?) {
        unRegisterShortcutsFn()

        val editor = e!!.getData(PlatformDataKeys.EDITOR)
        if (editor != null) actionPerformed(editor)
    }

    private fun actionPerformed(editor: Editor) {
        val keyToOffsetsMappings = getKeyToOffsetsMappings()

        if (keyToOffsetsMappings.size > 1) {
            val unHighlightCharsFn = highlightChars(editor, keyToOffsetsMappings)

            val nextActions: MutableList<AnAction> = ArrayList(keyToOffsetsMappings.size)
            val editorComponent = editor.getContentComponent()
            val unRegisterShortcutsFn = createUnRegisterShortcutsFn(editorComponent, nextActions)

            for ((char, offset) in keyToOffsetsMappings) {
                val cleanup = { cleanupFns.forEach { it() }; unRegisterShortcutsFn(); unHighlightCharsFn() }
                val next = MoveCaretToOffsetAction(keyToOffsetsMappings.get(char)!!, cleanup);
                next.registerCustomShortcutSet(CustomShortcutSet(createKeyStroke(char)), editorComponent)
                nextActions.add(next)
            }

            setStatusText(editor, "n.candidates.found", keyToOffsetsMappings.size)
            setupCancellation(
                    editorComponent,
                    *cleanupFns, unRegisterShortcutsFn, unHighlightCharsFn, createClearStatusFn(editor));
        } else {
            assert(keyToOffsetsMappings.size == 1)
            setStatusText(editor, "one.candidate");
            moveCaretTo(editor, keyToOffsetsMappings.values().first());
            runAll(*cleanupFns)
            unRegisterShortcutsFn()
        }
    }

    private fun getKeyToOffsetsMappings(): Map<Char, Int> {
        val affectedOffsets = offsets.take(Math.min(offsets.size, SUBSTITUTIONS.size))
        val mappings: MutableMap<Char, Int> = HashMap(offsets.size)
        var i = 0
        affectedOffsets.forEach { mappings.put(SUBSTITUTIONS[i++], it) }
        return mappings
    }
}
